codalab/apps/chahub/tasks.py

```diff
@@ -0,0 +1,221 @@
+        obj.chahub_needs_retry = True
+        obj.save(force_to_chahub=False)
+
+
+def batch_send_to_chahub(model, limit=None, retry_only=False):
+    qs = model.objects.all()
+    if retry_only:
+        qs = qs.filter(chahub_needs_retry=True)
+    if limit is not None:
+        qs = qs[:limit]
+
+    endpoint = model.get_chahub_endpoint()
+    data = [obj.get_chahub_data() for obj in qs if obj.get_chahub_is_valid()]
+    if not data:
+        logger.info('Nothing to send to Chahub at {}'.format(endpoint))
+        return
+    try:
+        logger.info("Sending all data to Chahub at {}".format(endpoint))
+        resp = _send(endpoint=endpoint, data=data)
+        logger.info("Response Status Code: {}".format(resp.status_code))
+        if resp.status_code != 201:
+            logger.warning('ChaHub Response Content: {}'.format(resp.content))
+    except ChahubException:
+        logger.info("There was a problem reaching Chahub. Retry again later")
+
+
+def chahub_is_up():
+    if not settings.CHAHUB_API_URL:
+        return False
+
+    logger.info("Checking whether ChaHub is online before sending retries")
+    try:
+        response = requests.get(settings.CHAHUB_API_URL)
+        if response.ok:
+            logger.info("ChaHub is online")
+            return True
+        else:
+            logger.info("Bad Status from ChaHub")
+            return False
+    except requests.exceptions.RequestException:
+        # This base exception works for HTTP errors, Connection errors, etc.
+        logger.info("Request Exception trying to access ChaHub")
+        return False
+
+
+def get_chahub_models():
+    return inheritors(ChaHubSaveMixin)
+
+
+@task(queue='site-worker')
+def do_chahub_retries(limit=None):
+    if not chahub_is_up():
+        return
+    chahub_models = get_chahub_models()
+    logger.info('Retrying for ChaHub models: {}'.format(chahub_models))
+
+    chahub_models = inheritors(ChaHubSaveMixin)
+    for model in chahub_models:
+        needs_retry = model.objects.filter(chahub_needs_retry=True)
+        if limit:
+            needs_retry = needs_retry[:limit]
+        for instance in needs_retry:
+            # Saving forces chahub update
+            instance.save(force_to_chahub=True)
+
+        # then delete all objects that need to be deleted
+        objects_to_be_deleted = model.objects.all_objects().filter(
+            chahub_timestamp__isnull=False,
+            chahub_needs_retry=False,
+            deleted=True
+        )
+        if limit is not None:
+            objects_to_be_deleted = objects_to_be_deleted[:limit]
+        for obj in objects_to_be_deleted:
+            obj.delete(real_delete=True)
+
+
+@task(queue='site-worker')
+def send_everything_to_chahub(limit=None):
+    if not chahub_is_up():
+        return
+    chahub_models = get_chahub_models()
+    for model in chahub_models:
+        batch_send_to_chahub(model, limit=limit)
+
+
+@task(queue='site-worker')
+def send_chahub_general_stats():
```

Comment:
Pretty sure none of this stuff here matters, since these statistics are calculated based on objects in the db on chahub, not as static counts anymore.... unless that isn't how we want chahub to work.

Reply:
cool will pull

Reply:
✔️